lib/context2d.js

```diff
@@ -18,6 +18,38 @@ const CanvasPattern = bindings.CanvasPattern
 const ImageData = bindings.ImageData
 const DOMMatrix = require('./DOMMatrix').DOMMatrix
 
+const proto = Context2d.prototype;
+
+const privateMethodNames = [
+  'setFont',
+  'setFillColor',
+  'setStrokeColor',
+  'setFillPattern',
+  'setStrokePattern',
+  'setTextBaseline',
+  'setTextAlignment',
+  'getMatrix',
+  'getFillColor',
+  'getStrokeColor',
+  'updateStateInfo',
+];
+
+const defaultProps = {
+  font: '10px sans-serif',
+  textBaseline: 'alphabetic',
+  textAlign: 'start',
+};
+
+const priv = privateMethodNames.reduce((methods, name) => {
+  const privName = `_${name}`;
+  const original = proto[privName];
+  methods[name] = original.call.bind(original);
+  delete proto[privName];
```

Comment:
Personally not a fan of modifying the prototype, `delete`ing props, and/or running too much logic when `require`ing a package, but I don't really have a better way to propose at the moment 😁 🤔

Reply:
That's basically what we have in #1247. However, deleting prototype properties is avoidable by exporting another class which will call the original static methods, but it will probably make all operations a bit slower.

Reply:
> but it will probably make all operations a bit slower.

I'm not sure if this is still true, but I remember V8 being very slow when modifying prototypes, and when using `delete` to remove properties 🤔

Reply:
Performed a test:

```js
console.time('Test');
for(let i = 0; i !== 1e6; i++){
  ctx.fillStyle += '';
  ctx.strokeStyle += '';
  ctx.globalAlpha |= 0;
  ctx.lineWidth |= 0;
  ctx.font += '';
}
console.timeEnd('Test');
```

Before PR: `3509.834ms`
After PR: `4650.146ms`

The extra `1100ms` is spent on WeakMap operations. Calls to old prototype methods doesn't seem to be slower than before.

> I remember V8 being very slow when modifying prototypes

I couldn't find anything about it, only found some articles on v8 blog saying that there was a bug in old version of v8 forcing internal objects' hash tables to be reallocated on deleting properties. Not sure if it is related, though. Currently, every lib file in node-canvas modifies a prototype of an internal exported class by defining getters and setters.

Also, from logical point of view, a prototype is just an ordinary object. Adding/deleting properties should be the same as defining a new object. Not sure how it can impact the performance. I performed another test: deleted random (unused) properties from the original implementation and ran the above script: time stays at `~3500ms`, so it seems that deleting prototype properties doesn't affect the performance.

I'm open for any ideas on how to improve the performance without exposing non-standard properties.

Reply:
I think it would be nice to just do all of this in C++ so that there's no clunky manipulation of the exported C++ class and so the code is all in one spot. `parseFont` will be a pain because that's a lot of string manipulation code to port, but maybe we could move as much as easily possible now and deal with the rest later. Things like `createLinearGradient`, `currentTransform` and `textAlign` are all trivial to port, and I'm happy to help with that effort if you guys agree.

> exporting another class which will call the original static methods, but it will probably make all operations a bit slower.

Invoking a static method with an instance of the object shouldn't be slow. You're talking about something like this?

```js
const {privates} = require("../build/Release/canvas.node")
Object.defineProperty(proto, 'fillStyle', {
  set(val) {
    privates.SetFillStyle(this, val)
  }
})
```

> Adding/deleting properties should be the same as defining a new object. Not sure how it can impact the performance.

On why performance is/can be affected:
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/The_performance_hazards_of__%5B%5BPrototype%5D%5D_mutation
* https://richardartoul.github.io/jekyll/update/2015/04/26/hidden-classes.html

What I don't know is when you're allowed to modify the proto without a penalty. ES5-ish code would set `Fn.prototype.method1 = ...`, then `method2 = ...`, etc. So is it parse-time that modification is allowed without penalty?

Reply:
(Sorry for dupe. GitHub posted my comment twice and then deleted both copies when I tried to delete one.)

I think it would be nice to just do all of this in C++ so that there's no clunky manipulation of the exported C++ class and so the code is all in one spot. `parseFont` will be a pain because that's a lot of string manipulation code to port, but maybe we could move as much as easily possible now and deal with the rest later. Things like `createLinearGradient`, `currentTransform` and `textAlign` are all trivial to port, and I'm happy to help with that effort if you guys agree.

> exporting another class which will call the original static methods, but it will probably make all operations a bit slower.

Invoking a static method with an instance of the object shouldn't be slow. You're talking about something like this?

```js
const {privates} = require("../build/Release/canvas.node")
Object.defineProperty(proto, 'fillStyle', {
  set(val) {
    privates.SetFillStyle(this, val)
  }
})
```

> Adding/deleting properties should be the same as defining a new object. Not sure how it can impact the performance.

On why performance is/can be affected:
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/The_performance_hazards_of__%5B%5BPrototype%5D%5D_mutation
* https://richardartoul.github.io/jekyll/update/2015/04/26/hidden-classes.html

What I don't know is when you're allowed to modify the proto without a penalty. ES5-ish code would set `Fn.prototype.method1 = ...`, then `method2 = ...`, etc. So is it parse-time that modification is allowed without penalty?

Reply:
> I think it would be nice to just do all of this in C++

That seems reasonable. Would something like [this](https://github.com/Hakerh400/node-canvas/commit/272a2ab71589051be2d57641846bf2f73bfbd8cc#diff-d52111b9458a13e196dab8299f822c29R1869) work?

> Things like `createLinearGradient`, `currentTransform` and `textAlign` are all trivial to port

The `currentTransform` rely on the JS implementation of `DOMMatrix`. Constructing it from C++ would probably be slow, wouldn't it?

> `parseFont` will be a pain

Can't we just cache the string user passed to the setter and return it back, like we did in JS land?

Reply:
> Would something like this work?

Yep!

> The currentTransform rely on the JS implementation of DOMMatrix. Constructing it from C++ would probably be slow, wouldn't it?

It's not too bad. I did something similar the other day by retrieving the `require` function and invoking it with the module/path. The alternative is calling some `init`-ish C++ function with the `DOMMatrix` constructor and storing a persistent reference to it. I doubt anyone is using `currentTransform` in very hot code, so I wouldn't worry much about the extra 10s of nanoseconds here.

> Can't we just cache the string user passed to the setter and return it back, like we did in JS land?

Yeah, but I meant porting parse-font.js would be a bit of a pain. The regexps should port directly to C++ though, so maybe not so bad.